### ISSUE TITLE
[Proposal] Deprecating "approximate histogram" in favor of new sketches

### ISSUE THREAD

**Issue by jon-wei** [apache/incubator-druid#6869]
Deprecating "approximate histogram" in favor of new sketches
==========================

Motivation
----------------
Druid's "approximate histogram" aggregator has several significant drawbacks:
- No formal error bounds
- Accuracy is heavily data dependent
  - Doesn't handle sorted data well
  - Doesn't handle long tails well

It's not uncommon for users to get bad results using that aggregator, without any clear idea as to why, e.g. https://github.com/apache/incubator-druid/issues/6853.

Druid now has better options for quantile/histogram approximations:
- The quantiles sketch from the DataSketches extension (http://druid.io/docs/latest/development/extensions-core/datasketches-quantiles.html) is a better choice, with formal error bounds and a distribution independent algorithm.
- The upcoming moments sketch aggregator (https://github.com/apache/incubator-druid/pull/6581) is another option, using a distribution dependent algorithm with better performance/accuracy characteristics than Druid's "approximate histogram".

Proposed Changes
----------------
- Add a doc page containing the following guidance:
  - Document the advantages of the sketch algorithms over "approximate histogram" to encourage users to transition
  - Provide advice on how to choose between the quantiles sketch or the moments sketch
- Update docs to replace any examples/recommendations of "approximate histogram" as needed, mark "approximate histogram" as deprecated
- In line with what's being discussed in https://github.com/apache/incubator-druid/issues/6814 re: APPROX_COUNT_DISTINCT:
  - Change the APPROX_QUANTILE Druid SQL function to use whatever aggregator type is stored in a segment, and use a default option when used on a numeric column
  - Add individual APPROX_QUANTILE_* functions for each quantile estimation option

Changed Interfaces
----------------
The behavior of APPROX_QUANTILE in Druid SQL would change as described above.

Migration
----------------
Is it possible/valid for the newer sketch aggs to be able to operate on an old "approximate histogram"? If so, this would make migration easier for users. 

If such migration is not possible, then users will need to reingest existing data, accept the discontinuity, or continue using the old aggregator.

Alternatives
----------------
We could try to make improvements to the "approximate histogram" aggregator, but I think there's little value in doing so since better alternatives already exist.

**Comment by gianm** [comment]
This proposal looks good to me. approxHistogram had a good run.

> Provide advice on how to choose between the quantiles sketch or the moments sketch

Possibly a loaded question, but, do you have thoughts on what kind of advice we'd include here?

> Is it possible/valid for the newer sketch aggs to be able to operate on an old "approximate histogram"? If so, this would make migration easier for users.

My guess is no, since the algorithms are all different.

> Change the APPROX_QUANTILE Druid SQL function to use whatever aggregator type is stored in a segment, and use a default option when used on a numeric column

This will require a bit of an enhancement to the SQL metadata layer. Check out RowSignature, in particular this part:

```java
          case COMPLEX:
            // Loses information about exactly what kind of complex column this is.
            type = Calcites.createSqlTypeWithNullability(typeFactory, SqlTypeName.OTHER, true);
            break;
```

For `APPROX_QUANTILE` to be able to behave differently based on the type of sketch stored in the underlying column, we'll need to figure out some way to encode that information in the type. Or some way to make it accessible to the `SqlAggregator` conversion out-of-band. Encoding it in the type somehow is preferable, though, if possible, because avoiding the need for out-of-band info is always nice. It would also make it easier to expose it through the INFORMATION_SCHEMA.

**Comment by leerho** [comment]
This proposal makes lots of sense and I’ll help where I can.  Note that I’m OOO until next week and my access to the internet may be limited.

**Comment by niketh** [comment]
Makes sense. Also a note on size of Approx Historgram vs size of Quantile Sketch would be nice.

**Comment by yurmix** [comment]
I assume we will also move DataSketches away from the "experimental features" sub-menu?
 believe this might be confusing to people as for whether the technology is ready to be adopted.

**Comment by fjy** [comment]
👍 on datasketches no longer experimental

**Comment by leerho** [comment]
FYI I just finished an accuracy study of the Approximate Histogram.

It is really bad!  You can find it [here](https://datasketches.github.io/docs/Quantiles/DruidApproxHistogramStudy.html).

**Comment by leerho** [comment]
Next up is the Moment Sketch ... :)

**Comment by jon-wei** [comment]
@leerho Great, thanks for the analysis!

**Comment by drcrallen** [comment]
http://mail-archives.apache.org/mod_mbox/incubator-general/201902.mbox/%3CCA%2BUaPnt%3DUvbLr_v-4%2BYbAmHsAM-GqQG%2Bb%3DgOw3BL3Cemj%2BOwSA%40mail.gmail.com%3E Looks like Datasketches is getting an incubation proposal (thanks @leerho !)
